doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+mode.
+
+#### `SSL_read`, `SSL_read_ex`, `SSL_peek`, `SSL_peek_ex`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | CS            |
+
+**Blocking Considerations:** Blocks until at least one byte is available or an
+error occurs if in blocking mode (including the peek functions).
+
+#### `SSL_write`, `SSL_write_ex`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | CS            |
+
+We have to implement all of the following modes:
+
+- `SSL_ENABLE_PARTIAL_WRITE` on or off
+- `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER` on or off
+- Blocking mode on or off
+
+**Blocking Considerations:** Blocks until libssl has accepted responsibility for
+(i.e., copied) all data provided, or an error occurs, if in blocking mode.
+
+#### `SSL_pending`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | CS            |
+
+#### `SSL_has_pending`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | CS            |
+
+**TBD.** Options:
+
+  - Semantics unchanged or approximated (essentially, `SSL_pending() || any RXE
+    queued || any URXE queued`).
+  - Change semantics to only determine the return value based on if there is
+    data in the stream receive buffer.
+
+#### `SSL_shutdown`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | CS            |
+
+Semantics unchanged on QUIC Connection SSL objects.
+
+Probable implementation:
+
+  - If the peer did not yet close the connection, send CONNECTION_CLOSE once
+    (best effort, tick), return 0.
+  - Otherwise, if teardown is complete, return 1.
+
+Semantics for future QUIC Stream SSL objects TBD, but:
+
+  - We should have a way for `SSL_shutdown` to only affect the stream object
+    and not the entire connection, so that applications can pass SSL objects for
+    an individual stream to parts of themselves which expect something
+    resembling traditional TCP stream and then call `SSL_shutdown`.
+
+    A reasonable design here would be to have `SSL_shutdown` on a QUIC stream
+    SSL object only shut down that stream. However this would mean
+    `SSL_shutdown` behaves differently on the default stream (i.e., the QUIC
+    connection SSL object) to other streams. Thus a new API should probably be
+    added explicitly for QUIC stream shutdown. `SSL_shutdown` on a QUIC stream
+    object will redirect to this function, or it can be used explicitly on the
+    QUIC connection object if it has a default stream bound.
```

Comment:
I actually already added a `SSL_shutdown_ex` function which will be pushed shortly, but for a different purpose.

I think the idea of making `SSL_shutdown` work like this would be incredibly confusing.